aip/cloud/25001.md

```diff
@@ -0,0 +1,66 @@
+---
+id: 25001
+state: draft
+created: TBD
+placement:
+  category: gcp
+---
+
+# Empty Objects
+
+Fields of type Message **must not** be empty. Empty messages do not have uniform
+
+## Alternatives
+
+The presence of empty Messages are sometimes used to indicate that a flag is enabled. Alternatives to empty Messages could include a single boolean field within the Message to signify enablement or an Enum rather than an empty Message.
+
+### Add Field to Message
+For example:
+```proto
+message EnableMyFeature {
+}
+```
+Could be modeled instead as:
+
+```proto
+message MyFeature {
+  bool enabled = 1;
+}
+```
+
```

Comment:
I don't remember seeing this as purely enablement, outside a oneof. Often it's "here's a message which might contain more information later" and its presence is indicative of the choice for that feature.

Reply:
Agreed, it's usually an object that could contain more information later. I think that's where adding the `enabled` flag within the object works well, as new fields can be added alongside that field as they become available.

One of the problems I'm trying to solve here is that there are many ways to implement behavior off of an empty message, none of which are clear to the user. To a user that is new to an API seeing an empty object documented in the REST docs is confusing, and it is unclear how or why to use such a field. 

Some of this confusion is contributed to by our generated documentation for empty objects like the following. [DLP primitive transformations](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates#DeidentifyTemplate.PrimitiveTransformation) have a field replaceWithInfoTypeConfig that is an empty object. [The generated documentation for this field](https://cloud.google.com/dlp/docs/reference/rest/v2/projects.deidentifyTemplates#DeidentifyTemplate.ReplaceWithInfoTypeConfig) seems broken to a viewer, but is technically correct.

I don't think that the documentation generation is to blame here though. This pattern is nonstandard and nonintuitive to users, which is why generated tools and docs end up with a sub-par experience.

Reply:
But `enabled` won't always be suitable, for multiple reasons:

- Depending on the context, it may be *invalid* to have a configuration which isn't enabled, because it's always *got* to be used to retrieve something. Basically you're asserting that there are *always* more valid scenarios than an empty message implies, and I don't believe that assertion is justified.
- Adding `enabled` requires more work for anyone populating a configuration; it's easy to forget to explicitly enable it, at which point it may be hard to work out what's wrong
- If it *is* valid for `enabled` to be false, then a user retrieving a configuration where it's false will still see an empty object - is that really any better than the current situation? (I'd argue it's actually worse, if they've only accidentally left off `enabled`.)

I agree that we can and should improve documentation - but I believe that's all that we *should* do here. I don't believe we should be *artificially* requiring an additional field.

As a side-note, I'm going to do some research on:

- How many (and which) messages are currently empty across all Google APIs
- How many (and which) messages have been empty but then become non-empty

That may provide some additional food for thought.